ndb_api/clone_helpers.go

```diff
@@ -0,0 +1,278 @@
+
+	sourceDatabase, _ := GetDatabaseById(ctx, ndb_client, database.GetCloneSourceDBId())
+	databaseType := GetDatabaseTypeFromEngine(sourceDatabase.Type)
+	// database.
+	// Fetch the required profiles for the database
+	profilesMap, err := ResolveProfiles(ctx, ndb_client, databaseType, database.GetProfileResolvers())
+	if err != nil {
+		log.Error(err, "Error occurred while getting required profiles", "database name", database.GetName(), "isClone", database.IsClone())
+		return
+	}
+	// Required for dbParameterProfileIdInstance in MSSQL action args
+	reqData[common.PROFILE_MAP_PARAM] = profilesMap
+
+	// Creating a provisioning request based on the database type
+	requestBody = &DatabaseCloneRequest{
+		Name:           database.GetName(),
+		Description:    database.GetDescription(),
+		CreateDbServer: true,
+		Clustered:      false,
+		NxClusterId:    database.GetClusterId(),
+		// SSHPublicKey populated by request appenders for non mssql dbs
+		DbServerId:               "",
+		DbServerClusterId:        "",
+		DbserverLogicalClusterId: "",
+		TimeMachineId:            sourceDatabase.TimeMachineId,
+		SnapshotId:               database.GetCloneSnapshotId(),
+		UserPitrTimestamp:        "",
+		TimeZone:                 database.GetTimeZone(),
+		LatestSnapshot:           false,
+		NodeCount:                1,
+		Nodes: []Node{
+			{
+				VmName:              database.GetName() + "_CLONE_VM",
+				ComputeProfileId:    profilesMap[common.PROFILE_TYPE_COMPUTE].Id,
+				NetworkProfileId:    profilesMap[common.PROFILE_TYPE_NETWORK].Id,
+				NewDbServerTimeZone: "",
+				NxClusterId:         database.GetClusterId(),
+				Properties:          make([]string, 0),
+			},
+		},
+		// Added by request appenders as per the engine
+		ActionArguments:            []ActionArgument{},
+		Tags:                       make([]interface{}, 0),
+		VmPassword:                 "",
+		ComputeProfileId:           profilesMap[common.PROFILE_TYPE_COMPUTE].Id,
+		NetworkProfileId:           profilesMap[common.PROFILE_TYPE_NETWORK].Id,
+		DatabaseParameterProfileId: profilesMap[common.PROFILE_TYPE_DATABASE_PARAMETER].Id,
+	}
+	// Appending request body based on database type
+	appender, err := GetRequestAppender(databaseType)
+	if err != nil {
+		log.Error(err, "Error while getting a request appender")
+		return
+	}
+	requestBody, err = appender.appendCloningRequest(requestBody, database, reqData)
+	if err != nil {
+		log.Error(err, "Error while appending clone request")
+		return
+	}
+
+	log.Info("Database Cloning", "requestBody", requestBody)
+	log.Info("Returning from ndb_api.GenerateCloningRequest", "database name", database.GetName())
+	return
+}
+
+func (a *MSSQLRequestAppender) appendCloningRequest(req *DatabaseCloneRequest, database DatabaseInterface, reqData map[string]interface{}) (*DatabaseCloneRequest, error) {
+	req.SSHPublicKey = reqData[common.NDB_PARAM_SSH_PUBLIC_KEY].(string)
+	vmName := req.Name
+	dbName := "DATABASE_NAME" // TODO
```

Comment:
what is this TODO? does that need to be removed?

Reply:
Good catch, addressing.